enhancements/network/ip-interface-selection.yaml

```diff
@@ -0,0 +1,278 @@
+gets routed where. However, this work should be able to serve as the basis
+for a broader multi-nic feature so we should avoid designs that would limit
+future work in this area.
+
+## Proposal
+
+The following is a possibly incomplete list of places that we do IP/interface
+selection. Ideally these should all use a single mechanism to do that so they
+are all consistent with each other.
+
+- Node IP (Kubelet and CRIO)
+- configure-ovs
+- resolv-prepender
+- Keepalived
+- Etcd
+
+In some cases (resolv-prepender) it may not matter if the IP selected is
+consistent with the other cases, but Node IP, configure-ovs, and Keepalived all
+need to match because their functionality depends on it. I'm less familiar with
+the requirements for Etcd, but it seems likely that should match as well. In
+general, it seems best if all IP selection logic comes from one place, whether
+it's strictly required or not.
+
+### Workflow Description
+
+At deployment time the cluster administrator will include a manifest that sets
+KUBELET_NODEIP_HINT appropriately. The nodeip-configuration service (which
+will now be set as a dependency for all other services that need IP/interface
+selection) will use that value to determine the desired IP and interface for
+all services on the node. It will write the results of this selection to a
+well-known location which the other services will consume. This way, we don't
+need to duplicate the selection logic to multiple places. It will happen once
+and be reused as necessary.
+
+#### Example
+
+- resolv-prepender has to run before any other node ip selection can take
+  place. Without resolv.conf populated the nodeip-configuration service cannot
+  pull the runtimecfg image.
+  - Note: This is not relevant for UPI deployments.
+- nodeip-configuration runs and selects one or more IPs. It writes them to
+  the Kubelet and CRIO configuration files (this is the existing behavior).
+- nodeip-configuration also writes the following files (new behavior):
+  - /run/nodeip-configuration/primary-ip
+  - /run/nodeip-configuration/ipv4
+  - /run/nodeip-configuration/ipv6
+- When configure-ovs runs, it looks for the primary-ip file and bridges the
+  interface associated with that IP. It may also look at the ipv4 and ipv6
+  addresses to determine which IP versions should be present on the bridge.
+  If the nodeip-configuration files are not found, the logic will remain as
+  it is today.
+- When keepalived runs it will read the IP from the primary-ip file and use
+  the associated interface for VRRP traffic.
+  - Note: When [dual stack VIP support](https://github.com/openshift/enhancements/blob/master/enhancements/network/on-prem-dual-stack-vips.md)
+    is implemented it will need to look at the appropriate ipvX file for each
+    VIP. Until then, it should always use the primary IP.
+
+#### Variation [optional]
+
+We may want to rename KUBELET_NODEIP_HINT to reflect the fact that it will now
+affect more than just Kubelet.
```

Comment:
I agree. The variable name should be updated to reflect its new "capabilities".

Reply:
Added this to open questions to highlight that we need to agree on a name.